source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -476,41 +476,87 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
-  return decodeAfterBuffer(data, DecodeType::READ);
+  return decodeAndBuffer(data, zk_filter_read_buffer_, DecodeType::READ);
 }
 
 Network::FilterStatus DecoderImpl::onWrite(Buffer::Instance& data) {
-  return decodeAfterBuffer(data, DecodeType::WRITE);
+  return decodeAndBuffer(data, zk_filter_write_buffer_, DecodeType::WRITE);
 }
 
-Network::FilterStatus DecoderImpl::decodeAfterBuffer(Buffer::Instance& data, DecodeType dtype) {
+Network::FilterStatus DecoderImpl::decodeAndBuffer(Buffer::Instance& data,
+                                                   Buffer::OwnedImpl& zk_filter_buffer,
+                                                   DecodeType dtype) {
+  uint32_t zk_filter_buffer_len = zk_filter_buffer.length();
+  uint32_t data_len = data.length();
+
+  if (zk_filter_buffer_len > 0) {
+    if (zk_filter_buffer_len + data_len > INT_LENGTH) {
+      // Prepend ZooKeeper filter buffer to current buffer.
+      data.prepend(zk_filter_buffer);
+      decodeAndBufferHelper(data, data_len, zk_filter_buffer, zk_filter_buffer_len, dtype);
+      // Drain the prepended ZooKeeper filter buffer.
+      data.drain(zk_filter_buffer_len);
+    } else {
+      zk_filter_buffer.add(data);
+    }
+  } else if (zk_filter_buffer_len == 0) {
+    decodeAndBufferHelper(data, data_len, zk_filter_buffer, zk_filter_buffer_len, dtype);
+  }
+  return Network::FilterStatus::Continue;
+}
+
+void DecoderImpl::decodeAndBufferHelper(Buffer::Instance& data, uint32_t data_len,
+                                        Buffer::OwnedImpl& zk_filter_buffer,
+                                        uint32_t zk_filter_buffer_len, DecodeType dtype) {
   uint64_t offset = 0;
-  int32_t len = 0;
+  uint32_t len = 0;
+  // Boolean to check whether there is at least one full packet in the buffer.
+  bool has_full_packets = false;
 
-  while (offset < data.length()) {
+  while (offset < data_len) {
     try {
       // Peek packet length.
       len = helper_.peekInt32(data, offset);
-      ensureMinLength(len, INT_LENGTH + XID_LENGTH);
+      if (dtype == DecodeType::READ) {
+        ensureMinLength(len, XID_LENGTH + INT_LENGTH);
+      } else if (dtype == DecodeType::WRITE) {
+        ensureMinLength(len, XID_LENGTH + ZXID_LENGTH + INT_LENGTH);
+      }
       ensureMaxLength(len);
       offset += len;
+      if (offset <= data_len) {
+        has_full_packets = true;
+      }
     } catch (const EnvoyException& e) {
       ENVOY_LOG(debug, "zookeeper_proxy: decoding exception {}", e.what());
       callbacks_.onDecodeError();
-      return Network::FilterStatus::Continue;
+      return;
     }
   }
 
-  if (offset == data.length()) {
+  if (offset == data_len) {
     decode(data, dtype);
-    return Network::FilterStatus::Continue;
+  } else if (offset > data_len) {
+    std::string temp_data;
+
+    if (has_full_packets) {
+      offset -= INT_LENGTH + len;
+      // Decode full packets.
+      data.copyOut(0, offset, temp_data.data());
```

Comment:
Looks like buffer peek will also copy out the data, so I use copyOut directly here.
> https://github.com/envoyproxy/envoy/blob/30ebb2cbceb417c1ba7d482c2f618e55bd0cd5bf/envoy/buffer/buffer.h#L373